src/d3a/models/market/two_sided_pay_as_clear.py

```diff
@@ -133,126 +133,99 @@ def match_offers_bids(self):
-                                  trade_rate=clearing_rate,
-                                  trade_bid_info=trade_bid_info,
-                                  buyer_origin=trade.buyer_origin)
-                return accepted_bids
-            elif bid_energy > offer.energy:
-                trade._replace(seller_origin=offer.seller_origin)
-                self.accept_offer(offer_or_id=offer,
-                                  buyer=trade.offer.buyer,
-                                  energy=offer.energy,
-                                  already_tracked=already_tracked,
-                                  trade_rate=clearing_rate,
-                                  trade_bid_info=trade_bid_info,
-                                  buyer_origin=trade.buyer_origin)
-                updated_bid = trade.offer
-                updated_bid._replace(energy=trade.offer.energy - offer.energy)
-                trade._replace(offer=updated_bid)
-                accepted_bids = [trade] + accepted_bids
-                return accepted_bids
-            elif bid_energy < offer.energy:
-                trade._replace(seller_origin=offer.seller_origin)
-                offer_trade = self.accept_offer(
-                    offer_or_id=offer,
-                    buyer=trade.offer.buyer,
-                    energy=bid_energy,
-                    already_tracked=already_tracked,
-                    trade_rate=clearing_rate,
-                    trade_bid_info=trade_bid_info,
-                    buyer_origin=trade.buyer_origin)
+    @classmethod
+    def _create_bid_offer_matchings(cls, clearing_energy, offer_list, bid_list):
+        # Return value, holds the bid-offer matches
+        bid_offer_matchings = []
+        # Keeps track of the residual energy from offers that have been matched once,
+        # in order for their energy to be correctly tracked on following bids
+        residual_offer_energy = {}
+        for bid in bid_list:
+            bid_energy = bid.energy
+            while bid_energy > FLOATING_POINT_TOLERANCE:
+                # Get the first offer from the list
+                offer = offer_list.pop(0)
+                # See if this offer has been matched with another bid beforehand.
+                # If it has, fetch the offer energy from the residual dict
+                # Otherwise, use offer energy as is.
+                offer_energy = residual_offer_energy.get(offer.id, offer.energy)
+                if offer_energy - bid_energy > FLOATING_POINT_TOLERANCE:
+                    # Bid energy completely covered by offer energy
+                    # Update the residual offer energy to take into account the matched offer
+                    residual_offer_energy[offer.id] = offer_energy - bid_energy
+                    # Place the offer at the front of the offer list to cover following bids
+                    # since the offer still has some energy left
+                    offer_list.insert(0, offer)
```

Comment:
Shouldn't this ´offer´ be updated with residual energy? In case if it is and I missed it, please point me there.

Reply:
The offer and bid objects are never updated here. Instead, a more verbose approach was followed, where the offer_energy is kept on a residual_offer_energy dict and is updated every time an offer gets accepted partially. You can see the updated energy in line 183